kube-controllers/pkg/controllers/node/ipam.go

```diff
@@ -486,57 +556,171 @@ func (c *ipamController) onBlockDeleted(key model.BlockKey) {
 	delete(c.allBlocks, blockCIDR)
 	delete(c.nodesByBlock, blockCIDR)
 	delete(c.emptyBlocks, blockCIDR)
+
+	// Transition from known or unknown pool association to nil.
+	pool := c.poolsByBlock[blockCIDR]
+	delete(c.blocksByPool[pool], blockCIDR)
+	delete(c.poolsByBlock, blockCIDR)
+}
+
+func (c *ipamController) onPoolUpdated(pool *apiv3.IPPool) {
+	_, poolNet, err := cnet.ParseCIDR(pool.Spec.CIDR)
+	if err != nil {
+		log.WithError(err).Warnf("Unable to parse CIDR for IP Pool %s", pool.Name)
+		return
+	}
+
+	// Blocks may not have a known association to an IP Pool. This can happen when a Pool gets deleted, or if block
+	// updates appear before their associated pool updates. Blocks lacking association to an IP Pool are grouped under
+	// "no_ippool", and we check for transitions from unknown to known pool association on pool update.
+	for block := range c.blocksByPool[unknownPoolLabel] {
```

Comment:
If I understand correctly, we only ever need to do this if the update is a _create_ event rather than updating a pool we already know about, right?

If so, we should probably put this within the `if c.poolCache[pool.Name] == nil` check below.

Reply:
Correct, since pool CIDRs and block sizes can't change, we can optimize further by only attempting to associate unknown blocks on pool creation. I've made this adjustment now